In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv("adult_with_headers.csv")

In [3]:
df.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [4]:
df.tail()

age      workclass  fnlwgt    education  education_num  \
32556   27        Private  257302   Assoc-acdm             12   
32557   40        Private  154374      HS-grad              9   
32558   58        Private  151910      HS-grad              9   
32559   22        Private  201490      HS-grad              9   
32560   52   Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation relationship    race      sex  \
32556   Married-civ-spouse        Tech-support         Wife   White   Female   
32557   Married-civ-spouse   Machine-op-inspct      Husband   White     Male   
32558              Widowed        Adm-clerical    Unmarried   White   Female   
32559        Never-married        Adm-clerical    Own-child   White     Male   
32560   Married-civ-spouse     Exec-managerial         Wife   White   Female   

       capital_gain  capital_loss  hours_per_week  native_country  income  
32556             0             0              38   United-States   <=50K  
32557             0             0              40   United-States    >50K  
32558             0             0              40   United-States   <=50K  
32559             0             0              20   United-States   <=50K  
32560         15024             0              40   United-States    >50K

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [6]:
df.describe()

age        fnlwgt  education_num  capital_gain  capital_loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours_per_week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [7]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

In [8]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()

In [9]:
num_cols=df.select_dtypes(include=np.number).columns
df[num_cols]=sc.fit_transform(df[num_cols])

In [10]:
df[num_cols]

age    fnlwgt  education_num  capital_gain  capital_loss  \
0      0.030671 -1.063611       1.134739      0.148453      -0.21666   
1      0.837109 -1.008707       1.134739     -0.145920      -0.21666   
2     -0.042642  0.245079      -0.420060     -0.145920      -0.21666   
3      1.057047  0.425801      -1.197459     -0.145920      -0.21666   
4     -0.775768  1.408176       1.134739     -0.145920      -0.21666   
...         ...       ...            ...           ...           ...   
32556 -0.849080  0.639741       0.746039     -0.145920      -0.21666   
32557  0.103983 -0.335433      -0.420060     -0.145920      -0.21666   
32558  1.423610 -0.358777      -0.420060     -0.145920      -0.21666   
32559 -1.215643  0.110960      -0.420060     -0.145920      -0.21666   
32560  0.983734  0.929893      -0.420060      1.888424      -0.21666   

       hours_per_week  
0           -0.035429  
1           -2.222153  
2           -0.035429  
3           -0.035429  
4           -0.035429  
...               ...  
32556       -0.197409  
32557       -0.035429  
32558       -0.035429  
32559       -1.655225  
32560       -0.035429  

[32561 rows x 6 columns]

In [11]:
from sklearn.preprocessing import MinMaxScaler
mm=MinMaxScaler()

In [12]:
df[num_cols]=mm.fit_transform(df[num_cols])

In [13]:
df[num_cols]

age    fnlwgt  education_num  capital_gain  capital_loss  \
0      0.301370  0.044302       0.800000      0.021740           0.0   
1      0.452055  0.048238       0.800000      0.000000           0.0   
2      0.287671  0.138113       0.533333      0.000000           0.0   
3      0.493151  0.151068       0.400000      0.000000           0.0   
4      0.150685  0.221488       0.800000      0.000000           0.0   
...         ...       ...            ...           ...           ...   
32556  0.136986  0.166404       0.733333      0.000000           0.0   
32557  0.315068  0.096500       0.533333      0.000000           0.0   
32558  0.561644  0.094827       0.533333      0.000000           0.0   
32559  0.068493  0.128499       0.533333      0.000000           0.0   
32560  0.479452  0.187203       0.533333      0.150242           0.0   

       hours_per_week  
0            0.397959  
1            0.122449  
2            0.397959  
3            0.397959  
4            0.397959  
...               ...  
32556        0.377551  
32557        0.397959  
32558        0.397959  
32559        0.193878  
32560        0.397959  

[32561 rows x 6 columns]

In [15]:
cat_cols=df.select_dtypes(include=object).columns

In [16]:
from sklearn.preprocessing import OneHotEncoder
ohe=OneHotEncoder()

In [17]:
ohe_col=ohe.fit_transform(df[cat_cols])

In [18]:
print(ohe_col.toarray)

<bound method _cs_matrix.toarray of <32561x104 sparse matrix of type '<class 'numpy.float64'>'
	with 293049 stored elements in Compressed Sparse Row format>>


In [19]:
ohe_df=pd.DataFrame(ohe_col.toarray(),columns=ohe.get_feature_names_out())

In [20]:
ohe_df

workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  \
0               0.0                     0.0                   0.0   
1               0.0                     0.0                   0.0   
2               0.0                     0.0                   0.0   
3               0.0                     0.0                   0.0   
4               0.0                     0.0                   0.0   
...             ...                     ...                   ...   
32556           0.0                     0.0                   0.0   
32557           0.0                     0.0                   0.0   
32558           0.0                     0.0                   0.0   
32559           0.0                     0.0                   0.0   
32560           0.0                     0.0                   0.0   

       workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  \
0                          0.0                 0.0                      0.0   
1                          0.0                 0.0                      0.0   
2                          0.0                 1.0                      0.0   
3                          0.0                 1.0                      0.0   
4                          0.0                 1.0                      0.0   
...                        ...                 ...                      ...   
32556                      0.0                 1.0                      0.0   
32557                      0.0                 1.0                      0.0   
32558                      0.0                 1.0                      0.0   
32559                      0.0                 1.0                      0.0   
32560                      0.0                 0.0                      1.0   

       workclass_ Self-emp-not-inc  workclass_ State-gov  \
0                              0.0                   1.0   
1                              1.0                   0.0   
2                              0.0                   0.0   
3                              0.0                   0.0   
4                              0.0                   0.0   
...                            ...                   ...   
32556                          0.0                   0.0   
32557                          0.0                   0.0   
32558                          0.0                   0.0   
32559                          0.0                   0.0   
32560                          0.0                   0.0   

       workclass_ Without-pay  education_ 10th  ...  native_country_ Scotland  \
0                         0.0              0.0  ...                       0.0   
1                         0.0              0.0  ...                       0.0   
2                         0.0              0.0  ...                       0.0   
3                         0.0              0.0  ...                       0.0   
4                         0.0              0.0  ...                       0.0   
...                       ...              ...  ...                       ...   
32556                     0.0              0.0  ...                       0.0   
32557                     0.0              0.0  ...                       0.0   
32558                     0.0              0.0  ...                       0.0   
32559                     0.0              0.0  ...                       0.0   
32560                     0.0              0.0  ...                       0.0   

       native_country_ South  native_country_ Taiwan  \
0                        0.0                     0.0   
1                        0.0                     0.0   
2                        0.0                     0.0   
3                        0.0                     0.0   
4                        0.0                     0.0   
...                      ...                     ...   
32556                    0.0                     0.0   
32557                    0.0                     0.0   
32558                    0.0   

In [21]:
cat_cols

Index(['workclass', 'education', 'marital_status', 'occupation',
       'relationship', 'race', 'sex', 'native_country', 'income'],
      dtype='object')

In [22]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'income'],
      dtype='object')

In [23]:
df['Profit'] = df['capital_gain'] - df['capital_loss']

In [24]:
df['capital_gain_log'] = np.log1p(df['capital_gain'])

In [25]:
df['age_hours_interaction'] = df['age'] * df['hours_per_week']

In [27]:
!pip install ppscore
from sklearn.ensemble import IsolationForest
import ppscore as pps

iso_forest = IsolationForest(contamination=0.1)
df['outlier'] = iso_forest.fit_predict(df[['age', 'hours_per_week', 'capital_gain']])

In [30]:
df_cleaned = df[df['outlier'] == 1]

In [31]:
pps_matrix = pps.matrix(df_cleaned)

In [32]:
correlation_matrix = df_cleaned.corr()

In [33]:
print("PPS Matrix:\n", pps_matrix.head())
print("Correlation Matrix:\n", correlation_matrix)

PPS Matrix:
      x              y   ppscore            case  is_valid_score  \
0  age            age  1.000000  predict_itself            True   
1  age      workclass  0.007197  classification            True   
2  age         fnlwgt  0.000000      regression            True   
3  age      education  0.044126  classification            True   
4  age  education_num  0.000000      regression            True   

                metric  baseline_score  model_score                     model  
0                 None        0.000000     1.000000                      None  
1          weighted F1        0.598030     0.600923  DecisionTreeClassifier()  
2  mean absolute error        0.051506     0.052232   DecisionTreeRegressor()  
3          weighted F1        0.192200     0.227845  DecisionTreeClassifier()  
4  mean absolute error        0.120587     0.124197   DecisionTreeRegressor()  
Correlation Matrix:
                             age    fnlwgt  education_num  capital_gain  \
age      

conclusion:from PPS
Age and workclass have a very low PPS score of 0.007. This indicates that age has little predictive power over the workclass.

Age and education have a slightly stronger but still low PPS score of 0.044. this suggests a weak, but somewhat more significant relationship between age and education.

conclusion:from correlation
The Correlation matrix highlights strong linear relationships in a few cases (like age with age-hours interaction and capital gain with capital gain log), suggesting that some features might be redundant or highly dependent on others.